integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/DropCacheEventListeners.scala

```diff
@@ -0,0 +1,135 @@
+
+object DropCachePreAggEventListener extends OperationEventListener {
+
+  val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+
+  /**
+   * Called on a specified event occurrence
+   *
+   * @param event
+   * @param operationContext
+   */
+  override protected def onEvent(event: Event, operationContext: OperationContext): Unit = {
+    event match {
+      case dropCacheEvent: DropCacheEvent =>
+        val carbonTable = dropCacheEvent.carbonTable
+        val sparkSession = dropCacheEvent.sparkSession
+        val internalCall = dropCacheEvent.internalCall
+        if (carbonTable.isChildDataMap && !internalCall) {
+          throw new UnsupportedOperationException("Operation not allowed on child table.")
+        }
+
+        if (carbonTable.hasDataMapSchema) {
+          val childrenSchemas = carbonTable.getTableInfo.getDataMapSchemaList.asScala
+            .filter(_.getRelationIdentifier != null)
+          for (childSchema <- childrenSchemas) {
+            val childTable =
+              CarbonEnv.getCarbonTable(
+                TableIdentifier(childSchema.getRelationIdentifier.getTableName,
+                  Some(childSchema.getRelationIdentifier.getDatabaseName)))(sparkSession)
+            try {
+              val dropCacheCommandForChildTable =
+                CarbonDropCacheCommand(
+                  TableIdentifier(childTable.getTableName, Some(childTable.getDatabaseName)),
+                  internalCall = true)
+              dropCacheCommandForChildTable.processMetadata(sparkSession)
+            }
+            catch {
+              case e: Exception =>
+                LOGGER.error(
+                  s"Clean cache for PreAgg table ${childTable.getTableName} failed.", e)
+            }
+          }
+        }
+    }
+  }
+}
+
+
+object DropCacheBloomEventListener extends OperationEventListener {
+
+  val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+
+  /**
+   * Called on a specified event occurrence
+   *
+   * @param event
+   * @param operationContext
+   */
+  override protected def onEvent(event: Event, operationContext: OperationContext): Unit = {
+    event match {
+      case dropCacheEvent: DropCacheEvent =>
+        val carbonTable = dropCacheEvent.carbonTable
+        val sparkSession = dropCacheEvent.sparkSession
+        val cache = CacheProvider.getInstance().getCarbonCache
+        val currentTableSizeMap = operationContext.getProperty(carbonTable.getTableUniqueName)
+          .asInstanceOf[mutable.Map[String, (String, Long, Long)]]
+
+        val datamaps = DataMapStoreManager.getInstance().getDataMapSchemasOfTable(carbonTable)
+          .asScala
+        val segments = CarbonDataMergerUtil
+          .getValidSegmentList(carbonTable.getAbsoluteTableIdentifier).asScala
+
+        datamaps.foreach {
+          datamap =>
+            try {
+              val shardPaths = segments.flatMap {
```

Comment:
Move bloom cache key generation logic to a method

Reply:
Done.